### PR TITLE
agent: Add support for parsing env_template configuration files

### PR DESCRIPTION
**Pull request by dhuckins**
This pull requests adds support for parsing `env_template` and top-level `exec` in a given agent configuration file. It is part of the larger effort to add environment variable support within Vault Agent ([VLT-253](https://docs.google.com/document/d/1HLTkPtxUTtW-wMkePWoHmZoGt9lLg2POslt1ranaWE8/edit#)).

## Example Config:

```hcl
auto_auth {
  method {
    type = "token_file"
    config {
      token_file_path = "/Users/avean/.vault-token"
    }
  }
}

vault {
  address = "http://localhost:8200"
}

env_template "TEST_PASSWORD" {
  contents             = "{{ with secret \"secret/data/test\" }}{{ .Data.data.password }}{{ end }}"
  error_on_missing_key = true
}
env_template "TEST_USER" {
  contents             = "{{ with secret \"secret/data/test\" }}{{ .Data.data.user }}{{ end }}"
  error_on_missing_key = true
}

exec {
  command                   = ["./my-app.sh"]
  restart_on_secret_changes = "always"
  restart_stop_signal       = "SIGTERM"
}
```

## Related PRs:

- #20530 
- #20569 
- #20628 

 

[VLT-253]: https://hashicorp.atlassian.net/browse/VLT-253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:290]: Don't forget to add this test :)

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:1096]: Nit: this name is a bit ambiguous since it's used far from declaration, maybe something like `exec` or `execConfig` would be better?

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:1124]: Agreed, enum would be better here.

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:1154]: Nit: similar to above, `template` or `templateConfig` would be better.

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:1177]:
```suggestion
		if numberOfKeys := len(item.Keys); numberOfKeys != 1 {
			return fmt.Errorf("expected one and only one environment variable name, got %d", numberOfKeys)
```

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:1186]:
```suggestion
			return fmt.Errorf("duplicate environment variable %q detected", name)
```

File: command/agent/config/config_test.go
Comment (averche) [inline command/agent/config/config_test.go:2137]: Please also check for non-default restart signal & restart-on-secret-change

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:178]: I'm also OK with adding Directory if it is easy to support with the existing exec wrapper library. Otherwise, I'd say this is a post-beta feature.

Comment (dhuckins) [reply command/agent/config/config.go:178]: what was the reason behind this again?

Comment (averche) [reply command/agent/config/config.go:178]: To be more in-line with the existing [`exec`](https://developer.hashicorp.com/vault/docs/agent/template#exec), docker command, and other tools.

Comment (dhuckins) [reply command/agent/config/config.go:178]: would rather not split it, does it give that much of a benefit?
the exec part is passed in here https://github.com/hashicorp/consul-template/blob/f2d85d7490fb6ddd1a0ac2b455a5d706bf648e0b/manager/runner.go#L1229-L1230 which splits it up anyway

File: command/agent/config/config.go
Comment (dhuckins) [reply command/agent/config/config.go:180]: yes! (will add tests)
we cheat and use consul-template https://github.com/hashicorp/consul-template/blob/f2d85d7490fb6ddd1a0ac2b455a5d706bf648e0b/signals/signals_unix.go#L18-L46

File: command/agent/config/config.go
Comment (dhuckins) [reply command/agent/config/config.go:178]: let me follow up in a subsequent PR

File: command/agent/config/config.go
Comment (dhuckins) [inline command/agent/config/config.go:290]: @averche changing this to a list removed the ability to override on merge, can you add this to the validation piece (if not part of validation already)

File: command/agent/config/config.go
Comment (dhuckins) [reply command/agent/config/config.go:1094]: its the same style as the other parse functions 🤷

File: command/agent/config/test-fixtures/config-env-templates-no-name.hcl
Comment (averche) [inline command/agent/config/test-fixtures/config-env-templates-no-name.hcl:1]: This file should be called `bad-config-...`, but I also have something similar in the validation PR, so feel free to drop it.

File: command/agent/config/config_test.go
Comment (averche) [inline command/agent/config/config_test.go:2210]: Could we add a test with an invalid signal as well, something like "SIG-DOES-NOT-EXIST"

**Review by averche — COMMENTED** [review]
LGTM with a few minor comments. We just need to wait for the consul-template release to merge this

File: command/agent/config/config.go
Comment (averche) [reply command/agent/config/config.go:290]: I will definitely validate that there are no duplicates in the config.

For the merge behavior, I believe we can still use [slices.ContainsFunc](https://pkg.go.dev/golang.org/x/exp/slices#ContainsFunc) and override if the value exists.

Comment (dhuckins) [reply command/agent/config/config.go:290]: eh, would rather not override. if there's dups, just validate and exit

Comment (averche) [reply command/agent/config/config.go:290]: In which contexts does merging get used? This should probably dictate whether we override or not.

CC: @VioletHynes

Comment (VioletHynes) [reply command/agent/config/config.go:290]: Merging is used when e.g. two config files are presented or a directory of config was presented, and we need to merge them to form a single config.

Make sure we follow the same behaviour of how other lists are merged! Here's how Templates are merged (we should use the same strategy, which it looks like we are doing):
```
	for _, l := range c.Templates {
		result.Templates = append(result.Templates, l)
	}
	for _, l := range c2.Templates {
		result.Templates = append(result.Templates, l)
	}

```

By the way, I've been holding off merging until this is flagged ready to review, but if you want earlier review let me know :)

Comment (dhuckins) [reply command/agent/config/config.go:290]: yup! that's how I did it. except for regular templates, uniqueness isn't a problem so appending to a list is OK.
for env vars, we are erroring out if there is a duplicate env var in the validation step

Comment (VioletHynes) [reply command/agent/config/config.go:290]: That makes sense! Validation is I think done (or should be done!) after the merge

File: command/agent/config/config.go
Comment (averche) [inline command/agent/config/config.go:179]:
```suggestion
	RestartStopSignal      os.Signal `hcl:"-" mapstructure:"restart_stop_signal"`
```

per internal discussion

File: command/agent/config/config.go
Comment (dhuckins) [inline command/agent/config/config.go:1118]: using `os.Interrupt` was intentional here

in this case, it maps to syscall.SIGINT

Comment (averche) [reply command/agent/config/config.go:1118]: IIRC, SIGTERM is a more conventional default for soft exit

Comment (averche) [reply command/agent/config/config.go:1118]: If [this stackoverflow](https://stackoverflow.com/a/4047975) is to be believed:
> SIGTERM and SIGKILL are intended for general purpose "terminate this process" requests. SIGTERM (by default) and SIGKILL (always) will cause process termination. SIGTERM may be caught by the process (e.g. so that it can do its own cleanup if it wants to), or even ignored completely; but SIGKILL cannot be caught or ignored.
>
> SIGINT and SIGQUIT are intended specifically for requests from the terminal: particular input characters can be assigned to generate these signals (depending on the terminal control settings).

**Review by averche — APPROVED** [review]
This LGTM but I've made a number of changes so could use another pair of eyes please :)

File: command/agent/config/config.go
Comment (VioletHynes) [inline command/agent/config/config.go:25]: Nit: delete this empty line

File: command/agent/config/config_test.go
Comment (VioletHynes) [inline command/agent/config/config_test.go:15]: Nit: remove this empty line

File: command/agent/config/test-fixtures/bad-config-env-templates-invalid-signal.hcl
Comment (VioletHynes) [inline command/agent/config/test-fixtures/bad-config-env-templates-invalid-signal.hcl:7]: Nit: having examples with environment variables like this in our config file test fixtures might imply that this expands, but it doesn't, so I think we should avoid it.

Maybe instead, use
`/home/username/.vault-token` or similar?

File: command/agent/config/test-fixtures/bad-config-env-templates-no-name.hcl
Comment (VioletHynes) [inline command/agent/config/test-fixtures/bad-config-env-templates-no-name.hcl:16]: I think it might be beneficial to have a text fixture that uses the `group` functionality we're adding as part of this initiative, if that has been implemented so far?

**Review by VioletHynes — APPROVED** [review]
Looks good! Approved, with some small comments. None of them are blocking, but all should be easy enough to resolve :)

File: command/agent/config/test-fixtures/bad-config-env-templates-no-name.hcl
Comment (dhuckins) [reply command/agent/config/test-fixtures/bad-config-env-templates-no-name.hcl:16]: it has not

File: go.mod
Comment (averche) [reply go.mod:72]: I don't think so. Usually we only do separate PR's for go.mod updates when tagging/releasing api/sdk package changes.

**Pull request merged**
